drivers/adc/mcp356x/mcp356x.c

```diff
@@ -0,0 +1,542 @@
+  uint8_t mux_reg;
+  bool differential;
+  uint8_t gain_bits;
+  uint8_t reference_bits;
+};
+
+struct mcp356x_data {
+  struct channel_map_entry channel_map[MAX_CHANNELS];
+  uint8_t enabled_channels_bitmap;
+
+  // Configrations. If these registers match we don't need to rewrite them
+  uint8_t config0;
+  uint8_t config1;
+  uint8_t config2;
+  uint8_t config3;
+  uint8_t mux;
+};
+
+struct mcp356x_config {
+  struct spi_dt_spec bus;
+  uint8_t device_addr; // Specified on chip packaging. Normally 1
+
+  enum OSR osr;
+  enum PRE prescale;
+  enum CLK_SEL clock;
+};
+int mcp_read_reg_8(const struct mcp356x_config *config, enum MCP_Reg reg,
+                   uint8_t *result);
+
+int mcp_read_reg_24(const struct mcp356x_config *config, enum MCP_Reg reg,
+                    uint32_t *data);
+int mcp_write_reg_8(const struct mcp356x_config *config, enum MCP_Reg reg,
+                    uint8_t data);
+
+int mcp_write_reg_24(const struct mcp356x_config *config, enum MCP_Reg reg,
+                     uint32_t data);
+
+static int mcp356x_read_channel(const struct device *dev,
+                                const struct adc_sequence *sequence) {
+  const struct mcp356x_config *config = dev->config;
+  struct mcp356x_data *data = dev->data;
+
+  int res = 0;
+  if (sequence->options != NULL) {
+    LOG_ERR("This driver does not support repeated sequencing");
+    return -1;
+  }
+  int32_t num_channels = __builtin_popcount(sequence->channels);
+
+  if (sequence->buffer_size < 4 * num_channels) {
+    LOG_ERR(
+        "MCP356x buffer must be at least 4 * # of channels read long. Reading "
+        "%d needs %d bytes. was supplied %d",
+        num_channels, 4 * num_channels, sequence->buffer_size);
+    return -1;
+  }
+
+  uint32_t *sample_array = sequence->buffer;
+  uint32_t sample_index = 0;
+
+  for (uint8_t i = 0; i < MAX_CHANNELS; i++) {
+    uint8_t channel_id = i;
+    if (((sequence->channels >> channel_id) & 0b1) == 0) {
+      continue;
+    }
+
+    struct channel_map_entry cfg = data->channel_map[channel_id];
+    uint8_t config0 = (data->config0 & 0b01111111) | (1 << 7);
+    if (config0 != data->config0) {
+      (void)mcp_write_reg_8(config, MCP_Reg_CONFIG0, config0);
+      data->config0 = config0;
+    }
+
+    uint8_t config2 = (data->config2 & 0b11000111) | (cfg.gain_bits << 3);
+    if (config2 != data->config2) {
+      (void)mcp_write_reg_8(config, MCP_Reg_CONFIG2, config2);
+      data->config2 = config2;
+    }
+
+    uint8_t mux_reg = data->channel_map[i].mux_reg;
+    if (mux_reg != data->mux) {
+      (void)mcp_write_reg_8(config, MCP_Reg_MUX, mux_reg);
+      data->mux = mux_reg;
+
+      // Wait for result to come in after ADC reset because of MUX change
+      // right one
+      k_usleep(10);
+    }
+
+    uint32_t val = 12345;
+    res = mcp_read_reg_24(config, MCP_Reg_ADCDATA, &val);
+    if (res != 0) {
+      return res;
+    }
+    sample_array[sample_index] = val; // TODO FIX THIS
```

Comment:
?

Reply:
Forgot to cleanup a comment. This is the fix